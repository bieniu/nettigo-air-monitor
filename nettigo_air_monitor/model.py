"""Type definitions for NAM."""

from dataclasses import dataclass

import aiohttp


@dataclass
class ConnectionOptions:
    """Options for NAM."""

    host: str
    username: str | None = None
    password: str | None = None
    auth: aiohttp.BasicAuth | None = None

    def __post_init__(self) -> None:
        """Call after initialization."""
        if self.username is not None:
            if self.password is None:
                raise ValueError("Supply both username and password")

            object.__setattr__(
                self, "auth", aiohttp.BasicAuth(self.username, self.password)
            )


@dataclass(frozen=True)
class NAMSensors:
    """Data class for NAM sensors."""

    bh1750_illuminance: float | None
    bme280_humidity: float | None
    bme280_pressure: float | None
    bme280_temperature: float | None
    bmp180_pressure: float | None
    bmp180_temperature: float | None
    bmp280_pressure: float | None
    bmp280_temperature: float | None
    dht22_humidity: float | None
    dht22_temperature: float | None
    ds18b20_temperature: float | None
    heca_humidity: float | None
    heca_temperature: float | None
    mhz14a_carbon_dioxide: float | None
    pms_caqi: int | None
    pms_caqi_level: str | None
    pms_p0: float | None
    pms_p1: float | None
    pms_p2: float | None
    sds011_caqi: int | None
    sds011_caqi_level: str | None
    sds011_p1: float | None
    sds011_p2: float | None
    sht3x_humidity: float | None
    sht3x_temperature: float | None
    signal: float | None
    sps30_caqi: int | None
    sps30_caqi_level: str | None
    sps30_p0: float | None
    sps30_p1: float | None
    sps30_p2: float | None
    sps30_p4: float | None
    uptime: int | None
