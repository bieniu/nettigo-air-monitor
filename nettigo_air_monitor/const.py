"""Constants for nettigo-air-monitor library."""

from typing import Final

from aiohttp.client import ClientTimeout

ATTR_CONFIG: Final[str] = "config"
ATTR_DATA: Final[str] = "data"
ATTR_GPS_HEIGHT: Final[str] = "gps_height"
ATTR_GPS_LAT: Final[str] = "gps_lat"
ATTR_GPS_LON: Final[str] = "gps_lon"
ATTR_OTA: Final[str] = "ota"
ATTR_RESTART: Final[str] = "restart"
ATTR_SENSOR_VALUES: Final[str] = "sensordatavalues"
ATTR_UPTIME: Final[str] = "uptime"
ATTR_VALUES: Final[str] = "values"

DEFAULT_TIMEOUT: Final[ClientTimeout] = ClientTimeout(total=5)

ENDPOINTS: Final[dict[str, str]] = {
    ATTR_CONFIG: "http://{host}/config.json",
    ATTR_DATA: "http://{host}/data.json",
    ATTR_OTA: "http://{host}/ota",
    ATTR_RESTART: "http://{host}/reset",
    ATTR_VALUES: "http://{host}/values",
}

RENAME_KEY_MAP: Final[list[tuple[str, str]]] = [
    ("bmp_pressure", "bmp180_pressure"),
    ("bmp_temperature", "bmp180_temperature"),
    ("conc_co2_ppm", "mhz14a_carbon_dioxide"),
    ("humidity", "dht22_humidity"),
    ("sds_p1", "sds011_p1"),
    ("sds_p2", "sds011_p2"),
    ("temperature", "dht22_temperature"),
    ("ambient_light", "bh1750_illuminance"),
]

IGNORE_KEYS = ("GPS_date", "GPS_time")

MAC_PATTERN: Final[str] = r"([0-9a-fA-F]{2}[:]){5}([0-9a-fA-F]{2})"
