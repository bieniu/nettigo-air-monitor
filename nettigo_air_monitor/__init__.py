"""Python wrapper for getting air quality data from Nettigo Air Monitor devices."""

from __future__ import annotations

import logging
import re
from http import HTTPStatus
from typing import Any

from aiohttp import ClientConnectorError, ClientResponseError, ClientSession
from aqipy import caqi_eu
from dacite import from_dict
from tenacity import (
    after_log,
    retry,
    retry_if_exception_type,
    stop_after_attempt,
    wait_incrementing,
)

from .const import (
    ATTR_CONFIG,
    ATTR_DATA,
    ATTR_GPS_HEIGHT,
    ATTR_GPS_LAT,
    ATTR_GPS_LON,
    ATTR_OTA,
    ATTR_RESTART,
    ATTR_UPTIME,
    ATTR_VALUES,
    DEFAULT_TIMEOUT,
    ENDPOINTS,
    IGNORE_KEYS,
    MAC_PATTERN,
    RENAME_KEY_MAP,
)
from .exceptions import (
    ApiError,
    AuthFailedError,
    CannotGetMacError,
    InvalidSensorDataError,
    NotRespondingError,
)
from .model import ConnectionOptions, NAMSensors

_LOGGER = logging.getLogger(__name__)


class NettigoAirMonitor:
    """Main class to perform Nettigo Air Monitor requests."""

    def __init__(self, session: ClientSession, options: ConnectionOptions) -> None:
        """Initialize."""
        self.host = options.host
        self._options = options
        self._session = session
        self._software_version: str | None = None
        self._update_errors: int = 0
        self._auth_enabled: bool = False
        self._latitude: float | None = None
        self._longitude: float | None = None
        self._altitude: float | None = None

    @classmethod
    async def create(
        cls, session: ClientSession, options: ConnectionOptions
    ) -> NettigoAirMonitor:
        """Create a new device instance."""
        instance = cls(session, options)
        await instance.initialize()
        return instance

    async def initialize(self) -> None:
        """Initialize."""
        _LOGGER.debug("Initializing device %s", self.host)

        try:
            config = await self.async_check_credentials()
        except AuthFailedError:
            self._auth_enabled = True
        else:
            self._auth_enabled = config["www_basicauth_enabled"]
            self._software_version = config.get("SOFTWARE_VERSION")

    @staticmethod
    def _construct_url(arg: str, **kwargs: str) -> str:
        """Construct Nettigo Air Monitor URL."""
        return ENDPOINTS[arg].format(**kwargs)

    @staticmethod
    def _parse_sensor_data(data: dict[Any, Any]) -> dict[str, int | float]:
        """Parse sensor data dict."""
        result = {
            item["value_type"].lower(): float(item["value"])
            for item in data
            if item["value_type"] not in IGNORE_KEYS
        }

        for key, value in result.items():
            if "pressure" in key and value is not None:
                result[key] = value / 100

        if result.get("ambient_light") == -1:
            result.pop("ambient_light")

        for old_key, new_key in RENAME_KEY_MAP:
            if result.get(old_key) is not None:
                result[new_key] = result.pop(old_key)

        return result

    async def _async_http_request(self, method: str, url: str) -> Any:
        """Retrieve data from the device."""
        try:
            _LOGGER.debug("Requesting %s, method: %s", url, method)
            resp = await self._session.request(
                method,
                url,
                raise_for_status=True,
                auth=self._options.auth,
                timeout=DEFAULT_TIMEOUT,
            )
        except ClientResponseError as error:
            if error.status == HTTPStatus.UNAUTHORIZED.value:
                raise AuthFailedError("Authorization has failed") from error
            raise ApiError(
                f"Invalid response from device {self.host}: {error.status}"
            ) from error
        except (TimeoutError, ClientConnectorError) as error:
            _LOGGER.info("Invalid response from device: %s", self.host)
            raise NotRespondingError(
                f"The device {self.host} is not responding"
            ) from error

        _LOGGER.debug("Data retrieved from %s, status: %s", self.host, resp.status)
        if resp.status != HTTPStatus.OK.value:
            raise ApiError(f"Invalid response from device {self.host}: {resp.status}")

        return resp

    @retry(
        retry=retry_if_exception_type(NotRespondingError),
        stop=stop_after_attempt(5),
        wait=wait_incrementing(start=5, increment=5),
        after=after_log(_LOGGER, logging.DEBUG),
    )
    async def async_update(self) -> NAMSensors:
        """Retrieve data from the device."""
        url = self._construct_url(ATTR_DATA, host=self.host)

        resp = await self._async_http_request("get", url)

        data = await resp.json()

        self._software_version = data["software_version"]

        try:
            sensors = self._parse_sensor_data(data["sensordatavalues"])
        except (TypeError, KeyError) as error:
            raise InvalidSensorDataError("Invalid sensor data") from error

        if ATTR_GPS_LAT in sensors and self._latitude is None:
            self._latitude = sensors.pop(ATTR_GPS_LAT)

        if ATTR_GPS_LON in sensors and self._longitude is None:
            self._longitude = sensors.pop(ATTR_GPS_LON)

        if ATTR_GPS_HEIGHT in sensors and self._altitude is None:
            self._altitude = sensors.pop(ATTR_GPS_HEIGHT)

        if ATTR_UPTIME in data:
            sensors[ATTR_UPTIME] = int(data[ATTR_UPTIME])

        for sensor in ("pms", "sds011", "sps30"):
            value, data = caqi_eu.get_caqi(
                pm10_1h=sensors.get(f"{sensor}_p1"),
                pm25_1h=sensors.get(f"{sensor}_p2"),
                with_level=True,
            )
            if value is not None and value > -1:
                sensors[f"{sensor}_caqi"] = value
                sensors[f"{sensor}_caqi_level"] = data["level"].replace(" ", "_")

        result: NAMSensors = from_dict(data_class=NAMSensors, data=sensors)

        return result

    async def async_get_mac_address(self) -> str:
        """Retrieve the device MAC address."""
        url = self._construct_url(ATTR_VALUES, host=self.host)

        try:
            resp = await self._async_http_request("get", url)
        except NotRespondingError as error:
            raise ApiError(error.status) from error

        data = await resp.text()

        if not (mac := re.search(MAC_PATTERN, data)):
            raise CannotGetMacError("Cannot get MAC address from device")

        return mac[0]

    async def async_check_credentials(self) -> Any:
        """Request config.json to check credentials."""
        url = self._construct_url(ATTR_CONFIG, host=self.host)

        try:
            resp = await self._async_http_request("get", url)
        except NotRespondingError as error:
            raise ApiError(error.status) from error

        return await resp.json()

    @property
    def software_version(self) -> str | None:
        """Return software version."""
        return self._software_version

    @property
    def auth_enabled(self) -> bool:
        """Return True if basic auth is enabled."""
        return self._auth_enabled

    async def async_restart(self) -> None:
        """Restart the device."""
        url = self._construct_url(ATTR_RESTART, host=self.host)

        try:
            await self._async_http_request("post", url)
        except NotRespondingError as error:
            raise ApiError(error.status) from error

    async def async_ota_update(self) -> None:
        """Trigger OTA update."""
        url = self._construct_url(ATTR_OTA, host=self.host)

        try:
            await self._async_http_request("post", url)
        except NotRespondingError as error:
            raise ApiError(error.status) from error

    @property
    def latitude(self) -> float | None:
        """Return latitude."""
        return self._latitude

    @property
    def longitude(self) -> float | None:
        """Return longitude."""
        return self._longitude

    @property
    def altitude(self) -> float | None:
        """Return altitude."""
        return self._altitude
