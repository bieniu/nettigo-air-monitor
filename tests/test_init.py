"""Tests for nettigo package."""

import json
from http import HTTPStatus
from typing import Any
from unittest.mock import Mock, patch

import pytest
from aiohttp import ClientConnectorError, ClientResponseError, ClientSession
from aioresponses import aioresponses
from syrupy import SnapshotAssertion
from tenacity import RetryError

from nettigo_air_monitor import (
    ApiError,
    AuthFailedError,
    CannotGetMacError,
    ConnectionOptions,
    InvalidSensorDataError,
    NettigoAirMonitor,
)

VALID_IP = "192.168.172.12"
INVALID_HOST = "http://nam.org"

VALUES = "MAC: AA:BB:CC:DD:EE:FF<br/>"

DATA_JSON_URL = "http://192.168.172.12/data.json"
CONFIG_JSON_URL = "http://192.168.172.12/config.json"
VALUES_URL = "http://192.168.172.12/values"


@pytest.mark.asyncio
async def test_valid_data(
    snapshot: SnapshotAssertion, valid_data: dict[str, Any]
) -> None:
    """Test with valid data."""
    session = ClientSession()
    options = ConnectionOptions(VALID_IP)

    with aioresponses() as session_mock:
        session_mock.get(CONFIG_JSON_URL, payload={"www_basicauth_enabled": False})
        session_mock.get(DATA_JSON_URL, payload=valid_data)
        session_mock.get(VALUES_URL, payload=VALUES)

        nam = await NettigoAirMonitor.create(session, options)
        mac = await nam.async_get_mac_address()
        sensors = await nam.async_update()

    await session.close()

    assert mac == "AA:BB:CC:DD:EE:FF"

    assert nam.software_version == "NAMF-2020-36"
    assert nam.latitude == 52.284921
    assert nam.longitude == 20.889263
    assert nam.altitude == 102.8
    assert sensors == snapshot


@pytest.mark.asyncio
async def test_caqi_value(snapshot: SnapshotAssertion) -> None:
    """Test CAQI value when PM10 and PM2.5 is None."""
    data = {"software_version": "NAMF-2020-36", "sensordatavalues": []}

    session = ClientSession()
    options = ConnectionOptions(VALID_IP)

    with aioresponses() as session_mock:
        session_mock.get(CONFIG_JSON_URL, payload={"www_basicauth_enabled": False})
        session_mock.get(DATA_JSON_URL, payload=data)
        session_mock.get(VALUES_URL, payload=VALUES)

        nam = await NettigoAirMonitor.create(session, options)
        sensors = await nam.async_update()

    await session.close()

    assert sensors == snapshot


@pytest.mark.asyncio
async def test_valid_data_with_auth(
    snapshot: SnapshotAssertion, valid_data: dict[str, Any]
) -> None:
    """Test with valid data with authorization."""
    session = ClientSession()
    options = ConnectionOptions(VALID_IP, "user", "pass")

    with aioresponses() as session_mock:
        session_mock.get(CONFIG_JSON_URL, payload={"www_basicauth_enabled": True})
        session_mock.get(DATA_JSON_URL, payload=valid_data)
        session_mock.get(VALUES_URL, payload=VALUES)

        nam = await NettigoAirMonitor.create(session, options)
        mac = await nam.async_get_mac_address()
        sensors = await nam.async_update()

    await session.close()

    assert mac == "AA:BB:CC:DD:EE:FF"

    assert nam.software_version == "NAMF-2020-36"
    assert nam.auth_enabled is True
    assert sensors == snapshot


@pytest.mark.asyncio
async def test_auth_failed() -> None:
    """Test auth failed."""
    session = ClientSession()
    options = ConnectionOptions(VALID_IP, "user", "pass")

    with aioresponses() as session_mock:
        session_mock.get(CONFIG_JSON_URL, payload={"www_basicauth_enabled": False})

        nam = await NettigoAirMonitor.create(session, options)

    with aioresponses() as session_mock:
        session_mock.post(
            "http://192.168.172.12/reset",
            exception=ClientResponseError(
                Mock(), Mock(), status=HTTPStatus.UNAUTHORIZED.value
            ),
        )

        with pytest.raises(AuthFailedError) as excinfo:
            await nam.async_restart()

    assert str(excinfo.value) == "Authorization has failed"

    await session.close()


@pytest.mark.asyncio
async def test_auth_enabled() -> None:
    """Test auth failed."""
    session = ClientSession()
    options = ConnectionOptions(VALID_IP, "user", "pass")

    with aioresponses() as session_mock:
        session_mock.get(
            CONFIG_JSON_URL,
            exception=ClientResponseError(
                Mock(), Mock(), status=HTTPStatus.UNAUTHORIZED.value
            ),
        )

        nam = await NettigoAirMonitor.create(session, options)

    await session.close()

    assert nam.auth_enabled is True


@pytest.mark.asyncio
async def test_http_404_code() -> None:
    """Test request ends with error."""
    session = ClientSession()
    options = ConnectionOptions(VALID_IP, "user", "pass")

    with aioresponses() as session_mock:
        session_mock.get(
            CONFIG_JSON_URL,
            exception=ClientResponseError(
                Mock(), Mock(), status=HTTPStatus.NOT_FOUND.value
            ),
        )
        with pytest.raises(ApiError) as excinfo:
            await NettigoAirMonitor.create(session, options)

    assert str(excinfo.value) == "Invalid response from device 192.168.172.12: 404"

    await session.close()


@pytest.mark.asyncio
async def test_api_error() -> None:
    """Test API error."""
    session = ClientSession()
    options = ConnectionOptions(VALID_IP)

    with aioresponses() as session_mock:
        session_mock.get(CONFIG_JSON_URL, payload={"www_basicauth_enabled": False})

        nam = await NettigoAirMonitor.create(session, options)

    with aioresponses() as session_mock:
        session_mock.get(DATA_JSON_URL, status=HTTPStatus.ACCEPTED.value)
        with pytest.raises(ApiError) as excinfo:
            await nam.async_update()

    assert str(excinfo.value) == "Invalid response from device 192.168.172.12: 202"

    await session.close()


@pytest.mark.asyncio
async def test_invalid_sensor_data() -> None:
    """Test InvalidSensorDataError."""
    with open("tests/fixtures/invalid_data.json", encoding="utf-8") as file:
        data = json.load(file)

    session = ClientSession()
    options = ConnectionOptions(VALID_IP)

    with aioresponses() as session_mock:
        session_mock.get(CONFIG_JSON_URL, payload={"www_basicauth_enabled": False})

        nam = await NettigoAirMonitor.create(session, options)

    with aioresponses() as session_mock:
        session_mock.get(DATA_JSON_URL, payload=data)

        with pytest.raises(InvalidSensorDataError) as excinfo:
            await nam.async_update()

    assert str(excinfo.value) == "Invalid sensor data"

    await session.close()


@pytest.mark.asyncio
async def test_cannot_get_mac() -> None:
    """Test CannotGetMacError error."""
    session = ClientSession()
    options = ConnectionOptions(VALID_IP)

    with aioresponses() as session_mock:
        session_mock.get(CONFIG_JSON_URL, payload={"www_basicauth_enabled": False})

        nam = await NettigoAirMonitor.create(session, options)

    with aioresponses() as session_mock:
        session_mock.get(VALUES_URL, payload="lorem ipsum")

        with pytest.raises(CannotGetMacError) as excinfo:
            await nam.async_get_mac_address()

    assert str(excinfo.value) == "Cannot get MAC address from device"

    await session.close()


@pytest.mark.asyncio
async def test_init_device_not_repond() -> None:
    """Test init when device is not responding."""
    session = ClientSession()
    options = ConnectionOptions(VALID_IP)

    with aioresponses() as session_mock:
        session_mock.get(CONFIG_JSON_URL, exception=TimeoutError(Mock(), Mock()))

        with pytest.raises(ApiError) as excinfo:
            await NettigoAirMonitor.create(session, options)

    assert str(excinfo.value) == "The device 192.168.172.12 is not responding"

    await session.close()


@pytest.mark.asyncio
async def test_get_ma_device_not_repond() -> None:
    """Test get_mac when device is not responding."""
    session = ClientSession()
    options = ConnectionOptions(VALID_IP)

    with aioresponses() as session_mock:
        session_mock.get(CONFIG_JSON_URL, payload={"www_basicauth_enabled": False})

        nam = await NettigoAirMonitor.create(session, options)

    with aioresponses() as session_mock:
        session_mock.get(VALUES_URL, exception=TimeoutError(Mock(), Mock()))

        with pytest.raises(ApiError) as excinfo:
            await nam.async_get_mac_address()

    await session.close()

    assert str(excinfo.value) == "The device 192.168.172.12 is not responding"


@pytest.mark.asyncio
async def test_username_without_password() -> None:
    """Test error when username is provided without password."""
    with pytest.raises(
        ValueError, match="Supply both username and password"
    ) as excinfo:
        ConnectionOptions(VALID_IP, "user")

    assert str(excinfo.value) == "Supply both username and password"


@pytest.mark.asyncio
@pytest.mark.parametrize(
    ("method", "endpoint"), [("async_restart", "reset"), ("async_ota_update", "ota")]
)
async def test_post_methods(method: str, endpoint: str) -> None:
    """Test post methods."""
    session = ClientSession()
    options = ConnectionOptions(VALID_IP)

    with aioresponses() as session_mock:
        session_mock.get(CONFIG_JSON_URL, payload={"www_basicauth_enabled": False})

        nam = await NettigoAirMonitor.create(session, options)

    with (
        aioresponses() as session_mock,
        patch(
            "nettigo_air_monitor.NettigoAirMonitor._async_http_request"
        ) as mock_request,
    ):
        session_mock.post(f"http://192.168.172.12/{endpoint}")

        method_to_call = getattr(nam, method)
        await method_to_call()

    await session.close()

    assert mock_request.call_count == 1
    assert mock_request.call_args[0][0] == "post"
    assert mock_request.call_args[0][1] == f"http://192.168.172.12/{endpoint}"


@pytest.mark.asyncio
@pytest.mark.parametrize(
    ("method", "endpoint"), [("async_restart", "reset"), ("async_ota_update", "ota")]
)
async def test_post_methods_fail(method: str, endpoint: str) -> None:
    """Test fail of the post methods."""
    session = ClientSession()
    options = ConnectionOptions(VALID_IP, "user", "pass")

    with aioresponses() as session_mock:
        session_mock.get(CONFIG_JSON_URL, payload={"www_basicauth_enabled": True})

        nam = await NettigoAirMonitor.create(session, options)

    with aioresponses() as session_mock:
        session_mock.post(
            f"http://192.168.172.12/{endpoint}", exception=TimeoutError(Mock(), Mock())
        )

        method_to_call = getattr(nam, method)
        with pytest.raises(ApiError) as excinfo:
            await method_to_call()

    assert str(excinfo.value) == "The device 192.168.172.12 is not responding"

    await session.close()


@pytest.mark.parametrize(
    "exc", [TimeoutError(Mock(), Mock()), ClientConnectorError(Mock(), Mock())]
)
@pytest.mark.asyncio
async def test_retry_success(valid_data: dict[str, Any], exc: Exception) -> None:
    """Test retry which succeeded."""
    session = ClientSession()
    options = ConnectionOptions(VALID_IP)

    with aioresponses() as session_mock, patch("asyncio.sleep") as sleep_mock:
        session_mock.get(CONFIG_JSON_URL, payload={"www_basicauth_enabled": False})
        session_mock.get(DATA_JSON_URL, exception=exc)
        session_mock.get(DATA_JSON_URL, exception=exc)
        session_mock.get(DATA_JSON_URL, exception=exc)
        session_mock.get(DATA_JSON_URL, exception=exc)
        session_mock.get(DATA_JSON_URL, payload=valid_data)

        nam = await NettigoAirMonitor.create(session, options)
        await nam.async_update()

    assert sleep_mock.call_count == 4
    assert sleep_mock.call_args_list[0][0][0] == 5
    assert sleep_mock.call_args_list[1][0][0] == 10
    assert sleep_mock.call_args_list[2][0][0] == 15
    assert sleep_mock.call_args_list[3][0][0] == 20

    await session.close()


@pytest.mark.parametrize(
    "exc", [TimeoutError(Mock(), Mock()), ClientConnectorError(Mock(), Mock())]
)
@pytest.mark.asyncio
async def test_retry_fail(exc: Exception) -> None:
    """Test retry that failed."""
    session = ClientSession()
    options = ConnectionOptions(VALID_IP)

    with aioresponses() as session_mock, patch("asyncio.sleep") as sleep_mock:
        session_mock.get(CONFIG_JSON_URL, payload={"www_basicauth_enabled": False})
        session_mock.get(DATA_JSON_URL, exception=exc)
        session_mock.get(DATA_JSON_URL, exception=exc)
        session_mock.get(DATA_JSON_URL, exception=exc)
        session_mock.get(DATA_JSON_URL, exception=exc)
        session_mock.get(DATA_JSON_URL, exception=exc)

        nam = await NettigoAirMonitor.create(session, options)
        with pytest.raises(RetryError) as excinfo:
            await nam.async_update()

    await session.close()

    assert sleep_mock.call_count == 4
    assert sleep_mock.call_args_list[0][0][0] == 5
    assert sleep_mock.call_args_list[1][0][0] == 10
    assert sleep_mock.call_args_list[2][0][0] == 15
    assert sleep_mock.call_args_list[3][0][0] == 20

    assert "RetryError" in str(excinfo.value)


@pytest.mark.asyncio
async def test_illuminance_wrong_value() -> None:
    """Test with wrong value for illuminance."""
    session = ClientSession()
    options = ConnectionOptions(VALID_IP)

    data = {
        "software_version": "NAMF-2020-36",
        "age": "144",
        "measurements": "285",
        "uptime": "45632",
        "sensordatavalues": [{"value_type": "ambient_light", "value": "-1"}],
    }

    with aioresponses() as session_mock:
        session_mock.get(CONFIG_JSON_URL, payload={"www_basicauth_enabled": False})
        session_mock.get(DATA_JSON_URL, payload=data)
        session_mock.get(VALUES_URL, payload=VALUES)

        nam = await NettigoAirMonitor.create(session, options)
        sensors = await nam.async_update()

    await session.close()

    assert sensors.bh1750_illuminance is None
