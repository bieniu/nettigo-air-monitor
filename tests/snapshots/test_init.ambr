# serializer version: 1
# name: test_caqi_value
  NAMSensors(bh1750_illuminance=None, bme280_humidity=None, bme280_pressure=None, bme280_temperature=None, bmp180_pressure=None, bmp180_temperature=None, bmp280_pressure=None, bmp280_temperature=None, dht22_humidity=None, dht22_temperature=None, ds18b20_temperature=None, heca_humidity=None, heca_temperature=None, mhz14a_carbon_dioxide=None, pms_caqi=None, pms_caqi_level=None, pms_p0=None, pms_p1=None, pms_p2=None, sds011_caqi=None, sds011_caqi_level=None, sds011_p1=None, sds011_p2=None, sht3x_humidity=None, sht3x_temperature=None, signal=None, sps30_caqi=None, sps30_caqi_level=None, sps30_p0=None, sps30_p1=None, sps30_p2=None, sps30_p4=None, uptime=None)
# ---
# name: test_luftdaten_info_firmware
  NAMSensors(bh1750_illuminance=None, bme280_humidity=38.65, bme280_pressure=864.7022000000001, bme280_temperature=22.02, bmp180_pressure=None, bmp180_temperature=None, bmp280_pressure=None, bmp280_temperature=None, dht22_humidity=None, dht22_temperature=None, ds18b20_temperature=None, heca_humidity=None, heca_temperature=None, mhz14a_carbon_dioxide=None, pms_caqi=None, pms_caqi_level=None, pms_p0=None, pms_p1=None, pms_p2=None, sds011_caqi=2, sds011_caqi_level='very_low', sds011_p1=2.08, sds011_p2=1.33, sht3x_humidity=None, sht3x_temperature=None, signal=-68.0, sps30_caqi=None, sps30_caqi_level=None, sps30_p0=None, sps30_p1=None, sps30_p2=None, sps30_p4=None, uptime=None)
# ---
# name: test_valid_data
  NAMSensors(bh1750_illuminance=65.0, bme280_humidity=85.28, bme280_pressure=989.2061, bme280_temperature=10.6, bmp180_pressure=996.784, bmp180_temperature=10.8, bmp280_pressure=1022.0117999999999, bmp280_temperature=5.56, dht22_humidity=46.23, dht22_temperature=6.26, ds18b20_temperature=12.86, heca_humidity=59.73, heca_temperature=15.05, mhz14a_carbon_dioxide=865.0, pms_caqi=19, pms_caqi_level='very_low', pms_p0=6.0, pms_p1=10.0, pms_p2=11.0, sds011_caqi=34, sds011_caqi_level='low', sds011_p1=22.7, sds011_p2=20.0, sht3x_humidity=34.69, sht3x_temperature=6.28, signal=-85.0, sps30_caqi=54, sps30_caqi_level='medium', sps30_p0=31.23, sps30_p1=21.23, sps30_p2=34.32, sps30_p4=24.72, uptime=45632)
# ---
# name: test_valid_data_with_auth
  NAMSensors(bh1750_illuminance=65.0, bme280_humidity=85.28, bme280_pressure=989.2061, bme280_temperature=10.6, bmp180_pressure=996.784, bmp180_temperature=10.8, bmp280_pressure=1022.0117999999999, bmp280_temperature=5.56, dht22_humidity=46.23, dht22_temperature=6.26, ds18b20_temperature=12.86, heca_humidity=59.73, heca_temperature=15.05, mhz14a_carbon_dioxide=865.0, pms_caqi=19, pms_caqi_level='very_low', pms_p0=6.0, pms_p1=10.0, pms_p2=11.0, sds011_caqi=34, sds011_caqi_level='low', sds011_p1=22.7, sds011_p2=20.0, sht3x_humidity=34.69, sht3x_temperature=6.28, signal=-85.0, sps30_caqi=54, sps30_caqi_level='medium', sps30_p0=31.23, sps30_p1=21.23, sps30_p2=34.32, sps30_p4=24.72, uptime=45632)
# ---
